src/node_report_utils.cc

```diff
@@ -7,15 +7,52 @@ using node::MallocedBuffer;
 
 static constexpr auto null = JSONWriter::Null{};
 
+// Utility function to format socket information.
+static void ReportEndpoint(uv_handle_t* h,
+                           struct sockaddr* addr,
+                           const char* name,
+                           JSONWriter* writer) {
+  if (addr == nullptr) {
+    writer->json_keyvalue(name, null);
+    return;
+  }
+
+  uv_getnameinfo_t endpoint;
+  char* host = nullptr;
+  char hostbuf[INET6_ADDRSTRLEN];
+  const int family = addr->sa_family;
+  const int port = ntohs(family == AF_INET ?
+                         reinterpret_cast<sockaddr_in*>(addr)->sin_port :
+                         reinterpret_cast<sockaddr_in6*>(addr)->sin6_port);
+
+  if (uv_getnameinfo(h->loop, &endpoint, nullptr, addr, NI_NUMERICSERV) == 0) {
+    host = endpoint.host;
+    DCHECK_EQ(port, std::stoi(endpoint.service));
+  } else {
+    const void* src = family == AF_INET ?
+                      static_cast<void*>(
+                        &(reinterpret_cast<sockaddr_in*>(addr)->sin_addr)) :
+                      static_cast<void*>(
+                        &(reinterpret_cast<sockaddr_in6*>(addr)->sin6_addr));
+    if (uv_inet_ntop(family, src, hostbuf, sizeof(hostbuf)) == 0) {
+      host = hostbuf;
+    }
+  }
+  writer->json_objectstart(name);
+  if (host != nullptr) {
+    writer->json_keyvalue("host", host);
```

Comment:
I believe it's extremely unlikely for `host` to be `nullptr` (both `uv_getnameinfo()` and `uv_inet_ntop()` would have to fail) here but in the case that it is I've opted to omit the `host` key. Let me know if you'd prefer to emit `host: null` instead (i.e. the `host` key is always written), or something else (`host: '<unable to determine>'`?).